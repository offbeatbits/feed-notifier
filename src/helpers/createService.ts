import { serviceMappings } from '../config';
import {
  ChannelName,
  ServiceSettings,
} from '../types';

export const createService = (
  type: ChannelName,
  settings: ServiceSettings,
) =>
  new (serviceMappings[type] as any)(settings); // oof
