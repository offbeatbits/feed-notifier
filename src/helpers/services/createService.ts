import { TwitterService, MastodonService } from "../../services";
import { serviceMappings } from "../../config";
import {
  ChannelName,
  ServiceClassSettings,
  MastodonSettings,
} from "../../types.d";

type ServiceInstance = TwitterService | MastodonService;

export const createService = (
  type: ChannelName,
  settings: ServiceClassSettings
) =>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (serviceMappings[type] as any)({
    // oof :(
    username: settings.username,
    host: (settings as MastodonSettings).host,
    settings: settings.client,
  }) as ServiceInstance;
