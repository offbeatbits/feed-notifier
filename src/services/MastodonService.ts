import generator, {
  MegalodonInterface,
} from 'megalodon';
import { OAuth2Service } from '../models';
import { MastodonServiceSettings } from '../types';

export class MastodonService extends OAuth2Service {
  username: string;

  host: string;

  private mastodonClient: MegalodonInterface;

  constructor({
    username,
    host,
    settings,
  }: MastodonServiceSettings) {
    super(settings);
    this.username = username;
    this.host = host;
    this.mastodonClient = generator(
      'mastodon',
      this.host,
      this.accessToken,
    );
  }

  async post(content: string) {
    const response = await this.mastodonClient.postStatus(content);
    return response.data.uri;
  }
}
