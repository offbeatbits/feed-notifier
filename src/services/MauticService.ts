/* eslint-disable class-methods-use-this */
import MauticConnector from 'node-mautic';
import { MauticServiceSettings } from '../types';

export class MauticService {
  url: string;

  username: string;

  password: string;

  private mauticClient: MauticConnector;

  constructor({
    url,
    username,
    password,
  }: MauticServiceSettings) {
    this.url = url;
    this.username = username;
    this.password = password;
    this.mauticClient = new MauticConnector({
      apiUrl: this.url,
      username: this.username,
      password: this.password,
      timeoutInSeconds: 20,
    });
  }

  async post(/* content: string */) {
    console.log(this.mauticClient); // eslint-disable-line
    return 'mautic service response';
  }
}
