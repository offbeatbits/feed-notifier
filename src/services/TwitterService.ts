/* eslint-disable class-methods-use-this */
import { TwitterClient } from "twitter-api-client";
import { OAuth2Service } from "../models";
import { TwitterServiceSettings } from "../types.d";

export class TwitterService extends OAuth2Service {
  username: string;

  private twitterClient: TwitterClient;

  constructor({ username, settings }: TwitterServiceSettings) {
    super(settings);
    this.username = username;
    this.twitterClient = new TwitterClient({
      apiKey: this.key,
      apiSecret: this.secret,
      accessToken: this.accessToken,
      accessTokenSecret: this.accessTokenSecret,
    });
  }

  async post(content: string) {
    const response = await this.twitterClient.tweets.statusesUpdate({
      status: content,
    });
    return `https://twitter.com/${this.username}/status/${response.id_str}`;
  }
}
