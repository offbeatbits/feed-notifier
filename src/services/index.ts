export { MastodonService } from './MastodonService';
export { TwitterService } from './TwitterService';
