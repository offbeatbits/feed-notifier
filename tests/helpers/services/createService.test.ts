import { createService } from "../../../src/helpers/services/createService";

describe("createService", () => {
  const services = [
    {
      type: "twitter",
      settings: {
        username: "sample_twitter_username",
        client: {
          key: "sample_twitter_key",
          secret: "sample_twitter_secret",
          accessToken: "sample_twitter_access_token",
          accessTokenSecret: "sample_twitter_access_token_secret",
        },
      },
    },
    {
      type: "mastodon",
      settings: {
        username: "sample_mastodon_username",
        host: "sample_mastodon_host",
        client: {
          key: "sample_mastodon_key",
          secret: "sample_mastodon_secret",
          accessToken: "sample_mastodon_access_token",
        },
      },
    },
  ];

  services.forEach((service) => {
    it(`should create ${service.type} service`, () => {
      expect(() =>
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        createService(service.type as any, service.settings as any)
      ).not.toThrow();
    });
  });
});
