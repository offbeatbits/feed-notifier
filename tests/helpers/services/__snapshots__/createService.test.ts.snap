// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createService should create mastodon service 1`] = `
MastodonService {
  "accessToken": "sample_mastodon_access_token",
  "accessTokenSecret": undefined,
  "host": "sample_mastodon_host",
  "key": "sample_mastodon_key",
  "mastodonClient": Mastodon {
    "baseUrl": "sample_mastodon_host",
    "client": Client {
      "accessToken": "sample_mastodon_access_token",
      "baseUrl": "sample_mastodon_host",
      "cancelTokenSource": Object {
        "cancel": [Function],
        "token": CancelToken {
          "promise": Promise {
            "catch": [Function],
            "then": [Function],
          },
          "throwIfRequested": [Function],
        },
      },
      "proxyConfig": false,
      "userAgent": "megalodon",
    },
  },
  "secret": "sample_mastodon_secret",
  "username": "sample_mastodon_username",
}
`;

exports[`createService should create twitter service 1`] = `
TwitterService {
  "accessToken": "sample_twitter_access_token",
  "accessTokenSecret": "sample_twitter_access_token_secret",
  "key": "sample_twitter_key",
  "secret": "sample_twitter_secret",
  "twitterClient": TwitterClient {
    "transport": Transport {
      "cache": Cache {
        "cache": Map {},
        "maxByteSize": 16000000,
        "ttl": 360,
      },
      "credentials": Object {
        "accessToken": "sample_twitter_access_token",
        "accessTokenSecret": "sample_twitter_access_token_secret",
        "apiKey": "sample_twitter_key",
        "apiSecret": "sample_twitter_secret",
      },
      "oauth": Object {
        "_accessUrl": "https://api.twitter.com/oauth/access_token",
        "_authorize_callback": null,
        "_clientOptions": Object {
          "accessTokenHttpMethod": "POST",
          "followRedirects": true,
          "requestTokenHttpMethod": "POST",
        },
        "_consumerKey": "sample_twitter_key",
        "_consumerSecret": "sample_twitter_secret",
        "_defaultClientOptions": Object {
          "accessTokenHttpMethod": "POST",
          "followRedirects": true,
          "requestTokenHttpMethod": "POST",
        },
        "_headers": Object {
          "Accept": "*/*",
          "Connection": "close",
          "User-Agent": "Node authentication",
        },
        "_isEcho": false,
        "_nonceSize": 32,
        "_oauthParameterSeperator": ",",
        "_requestUrl": "https://api.twitter.com/oauth/request_token",
        "_signatureMethod": "HMAC-SHA1",
        "_version": "1.0A",
      },
    },
  },
  "username": "sample_twitter_username",
}
`;
